how to develop using docker (#2610)
PecanProject/pecan: DEV-INTRO.md

#### .env file
You can copy the [`env.example`](docker/env.example) file as .env in your pecan folder. The variables we want to modify are:

* `COMPOSE_PROJECT_NAME` set this to pecan, the prefix for all containers
* `PECAN_VERSION` set this to develop, the docker image we start with

#### folders

Next we will create the folders that will hold all the data for the docker containers using: `mkdir -p volumes/{lib,pecan,portainer,postgres,rabbitmq,traefik}`. The `volumes` folder will be ignored by git. You can create these at any location, however you will need to update the `docker-compose.dev.yml` file. The subfolders are used for the following:

- **lib** holds all the R packages for the specific version of PEcAn and R. This folder will be shared amongst all other containers, and will contain the compiled PEcAn code.
- **pecan** this holds all the data, such as workflows and any downloaded data.
- **portainer** if you enabled the portainer service this folder is used to hold persistent data for this service

> - **postgres** holds the actual database data. If you want to backup the database, you can stop the postgres container, zip up the folder.

Reviewer: maybe we can put a note that these volumes may get big and user choose location accordingly

Author: added some additional text

To compile the PEcAn code you can use the make command in either the rstudio container, or in the executor container. The script [`compile.sh`](sripts/compile.sh) will run make inside the executor container.

To make it easier to start the containers and not having to remember to use `docker-compose -f docker-compose.yml -f docker-compose.dev.yml` when you want to use the docker-compose comannd, you can rename `docker-compose.dev.yml` to `docker-compose.override.yml`. The docker-compose command will automatically use the `docker-compose.yml`, `docker-compose.override.yml` and the `.env` files to start the right containers with the correct parameters.

### Workflow Submission

You can submit your workflow either in the executor container or in rstudio container. For example to run the `docker.sipnet.xml` workflow located in the tests folder you can use: 

```

> docker-compose exec executor baseh

Reviewer: ```suggestion
docker-compose exec executor bash
```

# PEcAn Development

This is a minimal guide to getting started with PEcAn development under Docker. You can find more information about docker in the [pecan documentation](https://pecanproject.github.io/pecan-documentation/master/docker-index.html).

## Git Repository and Workflow

> The recommended workflow is gitflow which is described in the PEcAn documentation. For the repository we recommend using a [fork of the PEcAn repsitory](https://help.github.com/en/github/getting-started-with-github/fork-a-repo). In the `/scripts` folder there is a script called [syncgit.sh](scripts/syncgit.sh) that will help with synchronizing your fork with the official repository.

Reviewer 2: ```suggestion
We recommend following the the [gitflow](https://nvie.com/posts/a-successful-git-branching-model/) workflow and working in your own [fork of the PEcAn repsitory](https://help.github.com/en/github/getting-started-with-github/fork-a-repo). See the [PEcAn developer guide](book_source/02_demos_tutorials_workflows/05_developer_workflows/02_git/01_using-git.Rmd) for further details. In the `/scripts` folder there is a script called [syncgit.sh](scripts/syncgit.sh) that will help with synchronizing your fork with the official repository.
```

- **pecan** this holds all the data, such as workflows and any downloaded data.
- **portainer** if you enabled the portainer service this folder is used to hold persistent data for this service
- **postgres** holds the actual database data. If you want to backup the database, you can stop the postgres container, zip up the folder.
- **rabbitmq** holds persistent information of the message broker (rabbitmq). 
- **traefik** holds persisent data for the web proxy, that directs incoming traffic to the correct container.

These folders will hold all the persistent data for each of the respective containers and can grow. For example the postgres database is multiple GB. The pecan folder will hold all data produced by the workflows, including any downloaded data, and can grow to many giga bytes.

#### postgresql database

First we bring up postgresql (we will start RabbitMQ as well since it takes some time to start): `docker-compose -f docker-compose.yml -f docker-compose.dev.yml up -d postgres rabbitmq`. This will start postgresql and rabbitmq. We need to wait for a few minutes (you can look at the logs using `docker-compose logs postgres`) to see if it is ready.

> Once the database has finished starting up we will initialize the database using: `docker run --rm --network pecan_pecan pecan/db`. Once that is done we create two users for BETY:

Reviewer: just checking, does it matter if we get image with `latest` tag, or do we want to specify `pecan/db:develop` here?

Author: there actually is no pecan/db:develop. This is the image with the latest dump of the database, there is no code in this image. Keep in mind this should only be done once. (It will tell you that the database already exists and exit).

Reviewer: ok, makes sense

```
# guest user

> docker-compose run --rm bety user guestuser guestuser "Guest User" guestuser@example.com 4 4

Reviewer: for people who haven't renamed their dev.yml as override.yml, this command is missing arguments and cause
```
ERROR: Configuration for volume traefik specifies "o" driver_opt None, but a volume with the same name uses a different "o" driver_opt (bind). If you wish to use the new configuration, please remove the existing volume "pecan_traefik" first:
$ docker volume rm pecan_traefik
```
errors, it should be again something like `docker-compose -f docker-compose.yml -f docker-compose.dev.yml run --rm bety user guestuser guestuser "Guest User" guestuser@example.com 4 4`

Reviewer: although I think we shoul go the other way around, I mean let's not extend this command but instead rename `docker-compose.dev.yml` to `docker-compose.override.yml` in the beginning

Reviewer 3: @istfer what is the benefit of naming it `override.yml` instead of `dev.yml`?

Reviewer: I feel like it is less error prone for absolute beginners, the fewer stuff to type the better

# example user
docker-compose run --rm bety user carya illinois "Carya Demo User" carya@example.com 1 1
```

#### copy web config file

> The `docker-compose.dev.yaml` file has a section that will eanble editing the web application. This is by default commented out. If you want to uncoment it you will need to first copy the config.php from the docker/web folder. You can do this using `cp docker/web/config.docker.php web/config.php`.

Reviewer: not very clear if they must do this, or if this is just a suggestion..I think we won't use a stronger language we also need to explain what happens if they do/ if they don't

#### copy R packages

Next copy the R packages from a container to your local machine as the `volumes/lib` folder. This is not really needed, but will speed up the process of the first compilation. Later we will put our newly compiled code here as well. 

> You can copy all the data using `docker run -ti --rm -v ${PWD}/volumes/lib:/rlib pecan/base:develop cp -r /usr/local/lib/R/site-library/* /rlib/`. This will copy all compiled packages to your local machine.

Reviewer 3: @robkooper I'm getting an error from this step: `cp: cannot stat '/usr/local/lib/R/site-library/*': No such file or directory`. Which is accurate, as I only have a path `/usr/local/lib`. 

I also don't really understand what this step is doing, I think because it's mixing a Docker command with `cp`. Is `docker run` compiling the R packages or something?

Author: are you running that command inside docker, or on your local machine? docker run does not compile the files, however the pecan/base:develop image should have all the R files already compiled.

Reviewer 3: I think on my local machine? I'm running the command in my local Terminal from within my local pecan directory.

Author: make sure to run 
```
docker run -ti --rm -v ${PWD}/volumes/lib:/rlib pecan/base:develop cp -r /usr/local/lib/R/site-library/* /rlib/
```

It copies the data from the container to your local machine.

Reviewer 3: I think that's what I did, but that's what isn't working and I don't understand what it's doing to be able to troubleshoot: 
```
(base) Kristinas-MBP:pecan kristinariemer$ pwd
/Users/kristinariemer/Dropbox/Documents/UofA/Projects/PEcAn/pecan
(base) Kristinas-MBP:pecan kristinariemer$ docker run -ti --rm -v ${PWD}/volumes/lib:/rlib pecan/base:develop cp -r /usr/local/lib/R/site-library/* /rlib/
cp: cannot stat '/usr/local/lib/R/site-library/*': No such file or directory
```
Is `/usr/local/lib/R/site-library/` supposed to be a local folder?

Author: that command looks correct, and the folder `/usr/local/lib/R/site-library` is in the docker container.
can you do a `docker pull pecan/base:develop` and test one more time, while I debug it on my site as well.

Reviewer 3: Doing that docker pull did do some stuff and returned `Status: Downloaded newer image for pecan/base:develop`. But even after doing that, I'm still getting the same "no such file or directory" error from the `docker run`.

Author: ok this should work (fingers crossed). This will take a while
```
docker run -ti --rm -v ${PWD}/volumes/lib:/rlib pecan/base:develop cp -a /usr/local/lib/R/site-library/. /rlib/
```

Author: updated documentation as well.

Reviewer 3: This worked!

> You can copy all the data using `docker run -ti --rm -v ${PWD}/volumes/lib:/rlib pecan/base:develop cp -a /usr/local/lib/R/site-library/. /rlib/`. This will copy all compiled packages to your local machine.

Reviewer: I think somewhere around here first timers need to run `docker run -ti --rm --network pecan_pecan --volume pecan_pecan:/data --env FQDN=docker pecan/data:develop` to get core db files

Reviewer 3: I ran this command and also `docker pull pecan/db`, but it looks like I still don't have that met file with id = 5000000005. I get this error from the updated `docker.sipnet.xml`: 
```
2020-05-26 15:38:26 ERROR  [PEcAn.data.atmosphere::met.process] :
   Cannot process met without source information
Error in paste0("registration/register.", met, ".xml") :
  object 'met' not found
Calls: <Anonymous> -> <Anonymous> -> system.file -> file.path -> paste0
```
Somewhat relatedly, when trying to log into the localhost Betydb, using carya as the login and illinois as the password doesn't work for me. Is login info something else?

Reviewer: If `docker-compose run --rm bety user carya illinois "Carya Demo User" carya@example.com 1 1 ` returns success, that is the user-password combination yes..but this is one of the problematic lines, i.e. you need to do `docker-compose -f docker-compose.yml -f docker-compose.dev.yml run --rm bety user carya illinois "Carya Demo User" carya@example.com 1 1` if you don't have dev.yml as override.yml

In general, I seem to be missing all the records with 5000****  (assuming that's Winsconsin), I also don't have` http://localhost:8000/bety/formats/5000000002` for example which is the     `AMERIFLUX_BASE_HH` format

Reviewer 3: Thanks for suggesting this @istfer. I can't seem to get the bety user to work. I'm either getting the same traefik configuration error as above, or when I rename `docker-compose.dev.yml` to `docker-compose.override.yml`, it returns `/home/bety/docker/entrypoint.sh: line 86: exec: user: not found`. I'm not sure what I'm doing wrong?

Reviewer: Not sure, but if you are willing to start from scratch I have listed the steps here that worked for me (I'm working on ubuntu tho) https://github.com/robkooper/pecan/pull/5/files

Reviewer: but if you try it you need to run the code below after the 5th step since this branch is not merged yet:
```
git remote add robkooper https://github.com/robkooper/pecan.gitgit 
fetch robkooper
git checkout --track robkooper/docker-develop
```

Author: the record for the met data is from Wisconsin (5000000002). Not sure what the easiest way to clean this up. You can do the following (potentially not 100% tested).

```
docker-compose down
docker pull pecan/db
# not quite remove the old data, but effectively the same
mv volumes volumes.old
```

Now start the from "First time setup". This should load the data with the input record for [5000000002](http://localhost:8000/bety/inputs/5000000005).

Reviewer: Don't know about @KristinaRiemer but this still don't give me input or format record with id `5000000002`

Reviewer 3: That id didn't work for me a few days ago, I'll try these instructions again and let y'all know.

Reviewer 3: I went through @robkooper instructions above by running `docker-compose down`, `docker pull pecan/db`, and `mv volumes volumes.old` and then going through the instructions from "First time setup". I still don't see any data for 5000000002 or 5000000005. Do these come in with the "load example data" step? That's the only thing I couldn't get to work due to the FQDN error.

Reviewer: I tried `docker run --rm --network pecan_pecan --volume pecan_pecan:/data --env FQDN=docker pecan/data:develop` instead, but still no 5000000002

The compiled code is written to `/usr/local/lib/R/site-library` which is mapped to `volumes/lib` on your machine. This same folder is mounted in many other containers, allowing you to share the same PEcAn modules in all containers. Now if you change a module, and compile all other containers will see and use this new version of your module.

To compile the PEcAn code you can use the make command in either the rstudio container, or in the executor container. The script [`compile.sh`](sripts/compile.sh) will run make inside the executor container.

### Workflow Submission

You can submit your workflow either in the executor container or in rstudio container. For example to run the `docker.sipnet.xml` workflow located in the tests folder you can use: 

```
docker-compose exec executor bash
# inside the container
cd /pecan/tests

> R CMD ../web/workflow.R docker.sipnet.xml

Reviewer 3: I was able to get this workflow to run! It went until it hit the error below in the writing the config. 
```
2020-05-19 14:42:58 INFO   [write.config.SIPNET] :
   Writing SIPNET configs with input /home/carya/sites/niwot/niwot.clim
Error in gsub("@BINARY@", settings$model$binary, jobsh) :
  invalid 'replacement' argument
Calls: <Anonymous> ... <Anonymous> -> do.call -> write.config.SIPNET -> gsub
```
Is the intention for this sipnet test xml to work for anyone who's gotten the dev environment set up correctly?

Reviewer: When I looked at the given model id, there was no binary entry for the model..now I blew everything up I am not able to check again

but either way, the way we set up things -or the pecan code- fails to populate settings$model$binary, normally we shouldn't see this error..I mean I believe it is a docker-related error, but not sure what the exact source is

Author: this is not finished yet :( now i need to slowly debug to see what goes wrong.

Reviewer 3: @robkooper gotcha. Can you let us know when it's ready to test?

```

A better way of doing this is developed as part of GSOC, in which case you can leverage of the restful interface defined, or using the new R PEcAn API package.

# Directory Structure

Following are the main folders inside the pecan repository. 

### base (R packages)

These are the core packages of PEcAn. Most other packages will depend on the packages in this folder.

### models (R packages)

Each subfolder contains the required pieces to run the model in PEcAn

### modules (R packages)

Contains packages that either do analysis, or download and convert different data products.

### web (PHP + javascript)

The Pecan web application

### shiny (R + shiny)

Each subfolder is its own shiny application.

### book_source (RMarkdown)

The PEcAn documentation that is compiled and uploaded to the PEcAn webpage.

### docker

Some of the docker build files. The Dockerfiles for each model are placed in the models folder.

### scripts

Small scripts that are used as part of the development and installation of PEcAn.

# Advanced Development Options

## docker-compose

> To make it easier to start the containers and not having to remember to use `docker-compose -f docker-compose.yml -f docker-compose.dev.yml` when you want to use the docker-compose comannd, you can rename `docker-compose.dev.yml` to `docker-compose.override.yml`. The docker-compose command will automatically use the `docker-compose.yml`, `docker-compose.override.yml` and the `.env` files to start the right containers with the correct parameters.

Reviewer: I think it will be safer to put the renaming again to the beginning and get rid of all these `-f` flags

I think it will take someone longer to have issues because they forgot they had extra configurations in the override, I mean that feels like a bit more advanced level of developing than just running the stack

## Developing in Docker

If running on a linux system it is recommended to add your user to the docker group. This will prevent you from having to use `sudo` to start the docker containers, and makes sure that any file that is written to a mounted volume is owned by you. This can be done using `sudo adduser ${USER} docker`.

To get started with development in docker we need to bring up the docker stack first. In the main pecan folder you will find the [docker-compose.yml](docker-compose.yml) file that can be used to bring up the pecan stack. There is also the [docker-compose.dev.yaml](docker-compose.dev.yaml) file that adds additional containers, and changes some services to make it easier for development.

> By default docker-compose will use the files `docker-compose.yml` and `docker-compose.override.yml`. We will use the default `docker-compose.yml` file from PEcAn. The `docker-compose.override.yml` file can be used to configure it for your specific environment, in our case we will use it to setup the docker environment for development. To do this we copy the `docker-compose.dev.yml` file to `docker-compose.override.yml`. You can now use the command `docker-compose` to launch all the containers setup for develoopment.

Reviewer: ```suggestion
By default docker-compose will use the files `docker-compose.yml` and `docker-compose.override.yml`. We will use the default `docker-compose.yml` file from PEcAn. The `docker-compose.override.yml` file can be used to configure it for your specific environment, in our case we will use it to setup the docker environment for development. Copy the `docker-compose.dev.yml` file to `docker-compose.override.yml` to start working with your own override file, i.e. `cp docker-compose.dev.yml docker-compose.override.yml`. You can now use the command `docker-compose` to launch all the containers setup for development.
```

Reviewer: @KristinaRiemer  Rob now brought this section where he creates the override.yml up here, but maybe it was buried? How is the suggestion above? Instead of `mv` it might be better to use `cp` so that people don't commit the deletion of the dev.yml file by mistake.

Reviewer 3: Oh, duh, yeah that covers it. There's so much discussion in this PR that I'm having a hard time seeing and remembering everything! That looks good and make sense.

Author: I think this will be the PR with most comments.

> By default docker-compose will use the files `docker-compose.yml` and `docker-compose.override.yml`. We will use the default `docker-compose.yml` file from PEcAn. The `docker-compose.override.yml` file can be used to configure it for your specific environment, in our case we will use it to setup the docker environment for development. To do this we simply rename the `docker-compose.dev.yml` file to `docker-compose.override.yml`. You can now use the command `docker-compose` to launch all the containers setup for develoopment.

Reviewer: >we simply rename

I think it might be good to suggest using `cp` instead of `mv` otherwise they might accidentally commit deletion of the `docker-compose.dev.yml` file

If you in the past had loaded some of the data, but would like to start from scratch you can simply remove the `volumes` folder (and all the subfolders) and start with the "First time setup" section again.

### First time setup

The steps in this section only need to be done the fist time you start working with the stack in docker. After this is done you can skip these steps. You can find more detail about the docker commands in the [pecan documentation](https://pecanproject.github.io/pecan-documentation/master/docker-index.html).

Before doing anything it is recommended to make sure you have the lastest docker images ready. You can do a `docker-compose pull` to get the latest images.

* setup .env file
* create folders to hold the data
* load the postgresql database
* load some test data
* copy all R packages (optional but recommended) 
* setup for web folder development (optional)

#### .env file

You can copy the [`env.example`](docker/env.example) file as .env in your pecan folder. The variables we want to modify are:

* `COMPOSE_PROJECT_NAME` set this to pecan, the prefix for all containers
* `PECAN_VERSION` set this to develop, the docker image we start with

> 

Reviewer 3: ```suggestion
Both of these variables should also be uncommented by removing the # preceding them. 
```

> At the end you should see the following if you run the following command `egrep -v '^(#|$)' .env`

Reviewer: was `egrep -v '^(#|$)' .env` supposed to uncomment `COMPOSE_PROJECT_NAME` and `PECAN_VERSION` tags? because it didn't do anything for me.

Reviewer 3: I think this is just to check that the `.env` is set up correctly? Not to change anything in the file.

Reviewer: Oh I see, I misread

Reviewer 3: I think it should include explicit instructions to uncomment those lines though, so I added that as a suggestion.

- **lib** holds all the R packages for the specific version of PEcAn and R. This folder will be shared amongst all other containers, and will contain the compiled PEcAn code.
- **pecan** this holds all the data, such as workflows and any downloaded data.
- **portainer** if you enabled the portainer service this folder is used to hold persistent data for this service
- **postgres** holds the actual database data. If you want to backup the database, you can stop the postgres container, zip up the folder.
- **rabbitmq** holds persistent information of the message broker (rabbitmq). 
- **traefik** holds persisent data for the web proxy, that directs incoming traffic to the correct container.

These folders will hold all the persistent data for each of the respective containers and can grow. For example the postgres database is multiple GB. The pecan folder will hold all data produced by the workflows, including any downloaded data, and can grow to many giga bytes.

#### postgresql database

> First we bring up postgresql (we will start RabbitMQ as well since it takes some time to start): `docker-compose -f docker-compose.yml -f docker-compose.dev.yml up -d postgres rabbitmq`. This will start postgresql and rabbitmq. We need to wait for a few minutes (you can look at the logs using `docker-compose logs postgres`) to see if it is ready.

Reviewer: If they copied `docker-compose.dev.yml` to `docker-compose.override.yml` and have them both in the directory, this command (`docker-compose -f docker-compose.yml -f docker-compose.dev.yml up -d postgres rabbitmq`) will run just fine, but do we want to stick to `docker-compose up -d postgres rabbitmq`

Reviewer 3: I agree with @istfer on sticking with using `override.yml` throughout. Now that I'm running through the "First time setup" again to see if I can get that Wisconsin data, I'm having to switch back and forth between `dev.yml` and `override.yml`, which is confusing. 

I think it makes sense to have the file named `docker-compose.dev.yml` in the repo, but then at the beginning of these instructions (not sure exactly where) to have the command `mv docker-compose.dev.yml docker-compose.override.yml`. Then the rest of the documentation should be updated accordingly, which it looks like Istem has already been doing.

Author: yes, I forgot the update the documentation to fix this. Now you can just use `docker-compose up -d` no need to specify the different yml files.

> Once the database has finished starting up we will initialize the database. Before we run the container we want to make sure we have the latest database infomration, you can do this with `docker pull pecan/db`, which will make sure you have the latest version of the database ready. Now you can load the database using: `docker run --rm --network pecan_pecan pecan/db` (in this case we use the latest image instead of develop since it refers to the actual database data, and not the actual code). Once that is done we create two users for BETY:

Reviewer: ```suggestion
Once the database has finished starting up we will initialize the database. Before we run the container we want to make sure we have the latest database information, you can do this with `docker pull pecan/db`, which will make sure you have the latest version of the database ready. Now you can load the database using: `docker run --rm --network pecan_pecan pecan/db` (in this case we use the latest image instead of develop since it refers to the actual database data, and not the actual code). Once that is done we create two users for BETY:
```

Reviewer: this is just typo in "infomration" corrected

```
# guest user
docker-compose run --rm bety user guestuser guestuser "Guest User" guestuser@example.com 4 4

# example user
docker-compose run --rm bety user carya illinois "Carya Demo User" carya@example.com 1 1
```

#### load example data

> Once the database is loaded we can add some example data, some of the example runs and runs for the ED model, assume some of this data is available. To do this we first again make sure we have the latest code ready using `docker pull pecan/data:develop` and run this image using `docker run --rm --network pecan_pecan pecan/data:develop`. This can take some time, but all the data needed will be copied to the `/data` folder in the pecan containers (which is mounted from `volumes/pecan` in your current folder.

Reviewer: this is what I got with this command

```
docker run --rm --network pecan_pecan pecan/data:develop
NEED TO SET FQDN
```

> Once the database is loaded we can add some example data, some of the example runs and runs for the ED model, assume some of this data is available. To do this we first again make sure we have the latest code ready using `docker pull pecan/data:develop` and run this image using `docker run --rm --network pecan_pecan pecan/data:develop`. This can take some time, but all the data needed will be copied to the `/data` folder in the pecan containers (which is mounted from `volumes/pecan` in your current folder.

Reviewer: ```suggestion
Once the database is loaded we can add some example data, some of the example runs and runs for the ED model, assume some of this data is available. To do this we first again make sure we have the latest code ready using `docker pull pecan/data:develop` and run this image using `docker run --rm --network pecan_pecan --env FQDN=docker  pecan/data:develop`. This can take some time, but all the data needed will be copied to the `/data` folder in the pecan containers (which is mounted from `volumes/pecan` in your current folder.
```

> Once the database is loaded we can add some example data, some of the example runs and runs for the ED model, assume some of this data is available. To do this we first again make sure we have the latest code ready using `docker pull pecan/data:develop` and run this image using `docker run --rm --network pecan_pecan pecan/data:develop`. This can take some time, but all the data needed will be copied to the `/data` folder in the pecan containers (which is mounted from `volumes/pecan` in your current folder.

Reviewer: ```suggestion
Once the database is loaded we can add some example data, some of the example runs and runs for the ED model, assume some of this data is available. To do this we first again make sure we have the latest code ready using `docker pull pecan/data:develop` and run this image using `docker run -ti --rm --network pecan_pecan --volume pecan_pecan:/data --env FQDN=docker pecan/data:develop`. This can take some time, but all the data needed will be copied to the `/data` folder in the pecan containers (which is mounted from `volumes/pecan` in your current folder.
```

Reviewer: needs to be fixed, otherwise

```
docker run --rm --network pecan_pecan pecan/data:develop
NEED TO SET FQDN
```

Reviewer 3: I'm also getting this FQDN message here.

#### copy R packages (optional but recommended)

Next copy the R packages from a container to your local machine as the `volumes/lib` folder. This is not really needed, but will speed up the process of the first compilation. Later we will put our newly compiled code here as well. 

You can copy all the data using `docker run -ti --rm -v ${PWD}/volumes/lib:/rlib pecan/base:develop cp -a /usr/local/lib/R/site-library/. /rlib/`. This will copy all compiled packages to your local machine.

This only needs to be done once (or if the PEcAn base image changes drastically, for example a new version of R). You can also always delete all files in the `volumes/lib` folder, and recompile PEcAn from scratch.

#### copy web config file (optional)

> The `docker-compose.dev.yaml` file has a section that will enable editing the web application. This is by default commented out. If you want to uncoment it you will need to first copy the config.php from the docker/web folder. You can do this using `cp docker/web/config.docker.php web/config.php`.

Reviewer: ```suggestion
The `docker-compose.override.yml` file has a section that will enable editing the web application. This is by default commented out. If you want to uncoment it you will need to first copy the config.php from the docker/web folder. You can do this using `cp docker/web/config.docker.php web/config.php`.
```

### PEcAn Development

> To begin development we first have to bring up the full PEcAn stack. This assumes you have done once the steps above. You don't need to stop any running containers, you can use the following command to start all containers: `docker-compose -f docker-compose.yml -f docker-compose.dev.yml up -d`. At this point you have PEcAn running in docker.

Reviewer: likewise, it won't throw an error, but someone might have copied dev.yml to override.yml and made changes to override.yml at this point. shall we stick to `docker-compose up -d`

Author: will update the documentation, and just use docker-compose up command.

> The current folder (most likely your clone of the git repository) is mounted in some containers as `/pecan`, and in the case of rstudio also in your home folder as `pecan`. You can see which containers exactly in `docker-compose.dev.yaml`.

Reviewer: ```suggestion
The current folder (most likely your clone of the git repository) is mounted in some containers as `/pecan`, and in the case of rstudio also in your home folder as `pecan`. You can see which containers exactly in `docker-compose.override.yml`.
```

The steps in this section only need to be done the fist time you start working with the stack in docker. After this is done you can skip these steps. You can find more detail about the docker commands in the [pecan documentation](https://pecanproject.github.io/pecan-documentation/master/docker-index.html).

Before doing anything it is recommended to make sure you have the lastest docker images ready. You can do a `docker-compose pull` to get the latest images.

* setup .env file
* create folders to hold the data
* load the postgresql database
* load some test data
* copy all R packages (optional but recommended) 
* setup for web folder development (optional)

#### .env file

You can copy the [`env.example`](docker/env.example) file as .env in your pecan folder. The variables we want to modify are:

* `COMPOSE_PROJECT_NAME` set this to pecan, the prefix for all containers
* `PECAN_VERSION` set this to develop, the docker image we start with
Both of these variables should also be uncommented by removing the # preceding them. 
At the end you should see the following if you run the following command `egrep -v '^(#|$)' .env`

```
COMPOSE_PROJECT_NAME=pecan
PECAN_VERSION=develop
```

> #### folders

Reviewer 3: @robkooper  I'm going through these instructions from scratch to set up this Docker stack on a new server. What I'm actually supposed to _do_ here is now really confusing. From what I can tell, I'm supposed to: 
1. Run `mkdir -p $HOME/volumes/pecan/{lib,pecan,portainer,postgres,rabbitmq,traefik}`
2. Uncomment all of the lines after the `volumes:` line

Is this correct? Do I need to do anything else? 

I think this could be more clear by taking out all of the explanation about why it should be done this way. That content is useful but doesn't really belong here. It should mostly just tell me what to do, and I think the list of volumes with their purpose is also useful and should stay.

# PEcAn Development

> This is a minimal guide to getting started with PEcAn development.

Reviewer: I think it would help if we put a note here (e.g. "how to follow this document:") to tell people to read all doc first, and follow the steps we itemize at the end, like a true quickstart, e.g.:
```
1. git clone git@github.com/pecanproject/pecan
2. Changed directory to be in the pecan directory, cd pecan
3. Have docker-compose.override.yml and .env in the pecan folder
4. Change this in docker-compose.override.yml
5. Change that in .env
6. start database: docker-compose -p pecan up -d postgres
7. ...
```
 What I struggled before with our current documentation was that I wasn't sure where it wants me to take action (i.e. start typing the code) versus when it just tries to explain something to me. So I would suggest explicitly dividing this doc into first reading, then doing.

Reviewer 3: I agree with @ister, I think that would be a very useful way to structure this quickstart.

Author: now if only we got accepted in season of docs :(

Reviewer: OK, let's not have this hold this PR, once it is in, I can test it on my laptop and make this addition

## Git Repository and Workflow

> The recommended workflow is gitflow which is described in the PEcAn documentation. For the repository we recommend using a [fork of the PEcAn repsitory](https://help.github.com/en/github/getting-started-with-github/fork-a-repo). In the scripts folder there is a script called [syngit.sh](scripts/syncgit.sh) that will help with synchronizing your fork with the official repository.

Reviewer 4: ```suggestion
The recommended workflow is gitflow which is described in the PEcAn documentation. For the repository we recommend using a [fork of the PEcAn repsitory](https://help.github.com/en/github/getting-started-with-github/fork-a-repo). In the `/scripts` folder there is a script called [syncgit.sh](scripts/syncgit.sh) that will help with synchronizing your fork with the official repository.
```

## Developing in Docker

To get started with development in docker we need to bring up the docker stack first. In the main pecan folder you will find the [docker-compose.yml](docker-compose.yml) file that can be used to bring up the pecan stack. There is also the [docker-compose.dev.yaml](docker-compose.dev.yaml) file that adds additional containers, and changes some services to make it easier for development.

You can copy the `docker-compose.dev.yaml` to `docker-compose.override.yml`. Once that is done the `docker-compose` program will automatically use the `docker-compose.yml` and `docker-compose.override.yml` files, and you don't have to explicitly load them in the commands below.

### First time setup

You can copy the [`env.example`](docker/env.example) file as .env in your pecan folder. The variables we want to modify are:

* `COMPOSE_PROJECT_NAME` set this to pecan, the prefix for all containers
* `PECAN_VERSION` set this to develop, the docker image we start with

> Next we will create the folders that will hold all the data for the docker containers using: `mkdir -p volumes/{lib,pecan,portainer,postgres,rabbitmq,traefik}`. The `volumes` folder will be ignored by git.

Reviewer 5: As someone who hasn't make the transition from command line to docker, these parts are confusing. At a high level, (1) it's not clear why all this needs to be done and (2) if we need all this stuff why isn't it there already in the repo (e.g. why would we need to modify values in an example file, shouldn't the example just be set to the right thing? Does .env impact non-Docker installs, and if not why not just have .env there by default?) At a lower level, when the md is is rendered, will it be obvious where the env.example is located? Also, the volumes gets added in the pecan folder? (just trying to minimize places someone could mess up).

Also, it's not 100% clear to me whether the next set of steps using docker-composer are really 
"one-time" or things that need to be done every time you want to start up your pecan docker stack

Author: Added some more text to add some notes about the folders etc. The links in the document right now assume it is in the root of the pecan git repo, and point to the files in the git repo.

> Next we will create the folders that will hold all the data for the docker containers using: `mkdir -p volumes/{lib,pecan,portainer,postgres,rabbitmq,traefik}`. The `volumes` folder will be ignored by git.

Reviewer: could you elaborate why we need this? I didn't do any of these for my setup.

Author: will do (docker-compose.dev.yml sets up these folders).

Author: added some more text, let me know if this helps.

> First we bring up postgresql (we will start RabbitMQ as well since it takes some time to start): `docker-compose -f docker-compose.yml -f docker-compose.dev.yml up -d postgres rabbitmq`. This will start postgresql and rabbitmq. We need to wait for a few minutes (you can look at the logs using `docker-compose logs postgres`) to see if it is ready.

Reviewer: It is a bit confusing to use `docker-compose.dev.yml` here, but suggesting to copying it to  `docker.compose.override.yml` above. Although, I feel like for first time users this explicit command might be more helpful

Reviewer 4: I get the following error from this computer:

```
ERROR: Configuration for volume traefik specifies "o" driver_opt bind, but a volume with the same name uses a different "o" driver_opt (None). If you wish to use the new configuration, please remove the existing volume "pecan_traefik" first:
$ docker volume rm pecan_traefik
```

Reviewer 4: At this step I get errors when volumes (even stopped) exist

![image](https://user-images.githubusercontent.com/464871/81961768-af484100-95c7-11ea-95b1-ab1609ba63ae.png)

Author: looks like you have already the volumes called pecan_X, you can either remove them (`docker volume rm <name>`) or use a different prefix.

 **If you remove the volumes, all data will be lost!!**

To use a different prefix, change the `COMPOSE_PROJECT_NAME` in your .env

> During the time it takes for the database to start we will copy the R packages from a container to our `volumes/lib` folder. This folder contain the compiled R packages. Later we will put our newly compiled code here as well. You can copy all the data using `docker run -ti --rm -v ${PWD}/volumes/lib:/rlib pecan/base:develop cp -r /usr/local/lib/R/site-library/* /rlib/`. This will copy all compiled packages to your local machine. This only needs to be done once (or if the PEcAn base image changes drastically, for example a new version of R).

Reviewer 5: As a first timer, I can't figure out what's going on at this step (or why it needs to be done)

Reviewer: agreed with Mike, not clear for me as well. 

Why do we physically copy things instead of mounting? as mentioned above I didn't do any of these.. I guess I followed a different approach, though, I did git clone pecan and built the native version, and mounted my source code folder to docker container (per current documentation - [Local development and testing with docker](https://pecanproject.github.io/pecan-documentation/develop/docker-index.html#docker-local-devel)). I guess that's why I have all compiled packages on my local machine and I don't need to do this step? I think my approach defeats the purpose of not dealing with native pecan build, but now that I did it, I think I like my way better (for my purposes, I'm not saying it should be the way described on the documentation). But I'm also curious if what I did has disadvantages?

Author: updated documentation to clarify this. If you share your Rlib folder from your laptop and the docker images you might run into a case where the R version on your laptop and the R version in the docker containers differ, and the resulting compiled packages don't work.

Reviewer: I just have one more clarifying question, if I am determined to have a native pecan build in my machine at all times, do I need to worry about incompatibility about R version on my machine and the R version in the docker? I mean I would need to make all changes/updates on my local machine to have the native pecan build anyway right?

Reviewer: sounds good to me, thank you.

Author: as long as you are not sharing the R library folders, this should work. You can think of the docker containers as mini vm's. The best part about what you want to do is if the two versions of R are different you are testing the code at that point with different versions of R.

At this point postgresql should have started and is ready to load the database. This is done using `docker run -ti --rm --network pecan_pecan pecan/db`. This only needs to be done once (unless the volumes folder is removed).

Once the database is loaded we can bring up the rest of the docker stack using `docker-compose -f docker-compose.yml -f docker-compose.dev.yml up -d`. At this point you have PEcAn running in docker.

### PEcAn Development

> The current folder is mounted in some containers as `/pecan`. You can see which containers exactly in `docker-compose.dev.yaml`. You can now modify the code on your local machine, or you can use [rstudio](http://localhost:8000) in the docker stack. Once you made changes to the code you can compile the code either in the terminal of rstudio (`cd pecan && make`) or using `./scripts/compile.sh` from your machine (latter is nothing more than a shell script that runs `docker-compose exec executor sh -c 'cd /pecan && make'`. 

Reviewer: ```suggestion
The current folder is mounted in some containers as `/pecan`. You can see which containers exactly in `docker-compose.dev.yaml`. You can now modify the code on your local machine, or you can use [rstudio](http://localhost:8000/rstudio/) in the docker stack. Once you made changes to the code you can compile the code either in the terminal of rstudio (`cd pecan && make`) or using `./scripts/compile.sh` from your machine (latter is nothing more than a shell script that runs `docker-compose exec executor sh -c 'cd /pecan && make'`. 
```

Reviewer: I'm also curious if these will cause any permission errors would the code compiled in rstudio (1000:1000) be accessible by web (33:33)

Author: volumes are mounted from your laptop and will be owned by the user that is inside the container, in case of rstudio it will be user with id 1001 (rstudio) in case of executor it will be user with id 0 (root).

Not merging suggest since rstudio runs at` http://localhost:8000/rstudio/`

Reviewer: hmm I think I still have to worry about permission issues if I follow this correctly. On my laptop this is fine, but on the server I am not root, and when I run pecan I want my user to own those files. But again this doesn't need to hold this PR.

Also, I'm confused, my suggestion was to change the url from `http://localhost:8000` to `http://localhost:8000/rstudio/`? I also added a slash before `cd /pecan` for rstudio terminal command

Author: on the server you will need to add your user to the docker group `adduser istfer docker`

Reviewer: sounds good to me, thank you.

Author: added hint about adding yourself to docker group.

> If you submit new workflows through the webinterface it will use this new code in the executor, and any other containers that have `volumes/lib` mounted inside the containers (done by `docker-compose.dev.yaml`).

Reviewer: Hmm now does this `volumes` approach eliminate the need for rebuilding any of the images as you modify code? [Documentation](https://pecanproject.github.io/pecan-documentation/develop/docker-index.html#docker-build-images) says "rebuild pecan/base if you made changes to the PEcAn R package source code" or are these talking about different things?

Author: the compiled code is shared using the lib folder, so no need to build images since all containers use the same folder with packages.

Author: added this as well to documentation.

### Workflow Submission

A better way of doing this is developed as part of GSOC. You can leverage of the API folder (specifically submit_workflow.R).

# Directory Structure

Following are the main folders inside the pecan repository. 

### base (R packages)

These are the core packages of PEcAn. Most other packages will depend on the packages in this folder.

### models (R packages)

Each subfolder contains the required pieces to run the model in PEcAn

### modules (R packages)

> Contains packages that eitehr do analysis, or download and convert different data products.

Reviewer: ```suggestion
Contains packages that either do analysis, or download and convert different data products.
```

## Developing in Docker

> If running on a linux system it is recommended to add your user to the docker group. This will prevent you from having to use `sudo` to star the docker containers, and makes sure that any file that is written to a mounted volume is owned by you. This can be done using `sudo adduser ${USER} docker`.

Reviewer: ```suggestion
If running on a linux system it is recommended to add your user to the docker group. This will prevent you from having to use `sudo` to start the docker containers, and makes sure that any file that is written to a mounted volume is owned by you. This can be done using `sudo adduser ${USER} docker`.
```

If you in the past had loaded some of the data, but would like to start from scratch you can simply remove the `volumes` folder (and all the subfolders) and start with the "First time setup" section again.

### First time setup

The steps in this section only need to be done the fist time you start working with the stack in docker. After this is done you can skip these steps. You can find more detail about the docker commands in the [pecan documentation](https://pecanproject.github.io/pecan-documentation/master/docker-index.html).

Before doing anything it is recommended to make sure you have the lastest docker images ready. You can do a `docker-compose pull` to get the latest images.

* setup .env file
* create folders to hold the data
* load the postgresql database
* load some test data

> * copy all R packages (optional but recommended) 

Reviewer 3: Would this actually be optional for anyone working in a dev environment? For example, wouldn't one need to copy the R packages to `volumes` in order to run a workflow in the localhost RStudio?

Author: in theory it should work without it, just the problem is that the first time it will have to install ALL dependencies as well, so it will take a lot longer to compile PEcAn.

If you want to just run the workflow, there is no need for the development route as you can use the `docker.sipnet.xml` file and run that in the rstudio container.
